core/src/services/pcloud/backend.rs

```diff
@@ -0,0 +1,423 @@
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use async_trait::async_trait;
+use http::StatusCode;
+use log::debug;
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::fmt::Formatter;
+use std::sync::Arc;
+
+use super::core::parse_stat_metadata;
+use super::core::PcloudCore;
+use super::core::StatResponse;
+use super::error::parse_error;
+use super::error::PcloudError;
+use super::lister::PcloudLister;
+use super::writer::PcloudWriter;
+use super::writer::PcloudWriters;
+use crate::raw::*;
+use crate::*;
+
+/// Config for backblaze Pcloud services support.
+#[derive(Default, Deserialize)]
+#[serde(default)]
+#[non_exhaustive]
+pub struct PcloudConfig {
+    /// root of this backend.
+    ///
+    /// All operations will happen under this root.
+    pub root: Option<String>,
+    /// endpoint address of this backend.
+    pub endpoint: String,
+    /// username of this backend.
+    pub username: Option<String>,
+    /// password of this backend.
+    pub password: Option<String>,
+}
+
+impl Debug for PcloudConfig {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Config");
+
+        ds.field("root", &self.root);
+        ds.field("endpoint", &self.endpoint);
+        ds.field("username", &self.username);
+
+        ds.finish()
+    }
+}
+
+/// [pCloud](https://www.pcloud.com/) services support.
+#[doc = include_str!("docs.md")]
+#[derive(Default)]
+pub struct PcloudBuilder {
+    config: PcloudConfig,
+
+    http_client: Option<HttpClient>,
+}
+
+impl Debug for PcloudBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut d = f.debug_struct("PcloudBuilder");
+
+        d.field("config", &self.config);
+        d.finish_non_exhaustive()
+    }
+}
+
+impl PcloudBuilder {
+    /// Set root of this backend.
+    ///
+    /// All operations will happen under this root.
+    pub fn root(&mut self, root: &str) -> &mut Self {
+        self.config.root = if root.is_empty() {
+            None
+        } else {
+            Some(root.to_string())
+        };
+
+        self
+    }
+
+    /// endpoint of this backend.
+    /// https://api.pcloud.com for United States and https://eapi.pcloud.com for Europe
+    /// ref to [doc.pcloud.com](https://docs.pcloud.com/)
+    ///
+    /// It is required. e.g. `test`
+    pub fn endpoint(&mut self, endpoint: &str) -> &mut Self {
+        self.config.endpoint = endpoint.to_string();
+
+        self
+    }
+
+    /// username of this backend.
+    ///
+    /// It is required. e.g. `test`
```

Comment:
The example appears to lack purpose. Please clarify it. For instance, we could inform users about where to locate this or what is available on pCloud.

Reply:
fixed